lib/irb/pager.rb

```diff
@@ -47,12 +49,39 @@ def page(retain_content: false)
       rescue Errno::EPIPE
       end
 
-      private
-
       def should_page?
         IRB.conf[:USE_PAGER] && STDIN.tty? && (ENV.key?("TERM") && ENV["TERM"] != "dumb")
       end
 
+      def page_with_preview(width, height, modifier_proc)
+        out = PageOverflowIO.new(width, height) do |lines|
+          modified_output = modifier_proc.call(lines.join, true)
+          content, = take_first_page(width, height - 1) {|o| o.write modified_output }
+          content = content.chomp
+          content = "#{content}\e[0m" if Color.colorable?
+          $stdout.puts content
+        end
```

Comment:
nit: can we define the `overflow_callback` as a local? Like:

```rb
overflow_callback = proc do |lines|
  # ...
end
out = PageOverflowIO.new(width, height, overflow_callback)
```

Otherwise, I'd assume this block is called during the `PageOverflowIO`'s inititalization.